librz/analysis/arch/x86/x86_il.c

```diff
@@ -0,0 +1,3806 @@
+		goto lower_half;
+	}
+
+	/* Bit 15: Reserved,
+	always 1 on 8086 and 186,
+	always 0 on later models
+	Assuming 0 */
+	val = x86_bool_to_bv(IL_FALSE, size);
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(NT)), size));
+
+	/** Bit 12-13: IOPL,
+	I/O privilege level (286+ only),
+	always 1 on 8086 and 186
+	Assuming all 1 */
+	val = LOGOR(SHIFTL0(val, UN(size, 2)), UN(size, 0x3));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(OF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(DF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(IF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(TF)), size));
+
+lower_half:
+	if (size == 8) {
+		val = x86_bool_to_bv(VARG(EFLAGS(SF)), size);
+	} else {
+		val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(ZF)), size));
+	}
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(ZF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 2)), x86_bool_to_bv(VARG(EFLAGS(AF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 2)), x86_bool_to_bv(VARG(EFLAGS(PF)), size));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), UN(size, 1));
+	val = LOGOR(SHIFTL0(val, UN(size, 1)), x86_bool_to_bv(VARG(EFLAGS(CF)), size));
+
+	return val;
+}
+
+/**
+ * \brief Set the value of flags register
+ *
+ * \param val value to set the FLAGS register to
+ * \param size size of \p val
+ */
+RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+	RzILOpEffect *set_val = SETL("_flags", val);
+
+	RzILOpEffect *eff = SETG(EFLAGS(CF), LSB(VARL("_flags")));
+
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(2))));
+	eff = SEQ2(eff, SETG(EFLAGS(PF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(2))));
+	eff = SEQ2(eff, SETG(EFLAGS(AF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(2))));
+	eff = SEQ2(eff, SETG(EFLAGS(ZF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(1))));
+	eff = SEQ2(eff, SETG(EFLAGS(SF), LSB(VARL("_flags"))));
+
+	if (size == 8) {
+		return SEQ2(set_val, eff);
+	}
+
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(1))));
+	eff = SEQ2(eff, SETG(EFLAGS(TF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(1))));
+	eff = SEQ2(eff, SETG(EFLAGS(IF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(1))));
+	eff = SEQ2(eff, SETG(EFLAGS(DF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(1))));
+	eff = SEQ2(eff, SETG(EFLAGS(OF), LSB(VARL("_flags"))));
+	eff = SEQ2(eff, SETL("_flags", SHIFTR0(VARL("_flags"), U8(3))));
+	eff = SEQ2(eff, SETG(EFLAGS(NT), LSB(VARL("_flags"))));
+
+	/* Again, we will be ignoring bits over 16 and also ignore IOPL */
+
+	return SEQ2(set_val, eff);
+}
+
+/**
+ * ======== INSTRUCTION DOCUMENTATION FORMAT ========
```

Comment:
Maybe you could use `\@{` Doxygen paragraph grouping here?

Reply:
I don't exactly understand how I could use Doxygen grouping here. Also, this is more of a note, and I don't think it's a good idea to have it in the generated documentation anyways.

The Doxygen comments are only for the helper functions since anyone else can understand what the functions are doing and use them to extend the IL. Also, none of the functions here are exposed to the API, so again, I don't see any need for strictly following Doxygen, since this is more of an internal documentation.

Reply:
I'll wait for you reply before finalizing.

Reply:
Yes, fine for me.